servers/physics_2d/shape_2d_sw.h

```diff
@@ -48,7 +48,48 @@ SHAPE_CUSTOM, ///< Server-Implementation based custom shape, calling shape_creat
 
 class Shape2DSW;
 
-class ShapeOwner2DSW : public RID_Data {
+class ShapeOwner2DSW_Base : public RID_Data {
+public:
+	enum Type {
+		TYPE_AREA,
+		TYPE_BODY
+	};
+
+	_FORCE_INLINE_ Type get_type() const { return type; }
+	_FORCE_INLINE_ bool is_static() const { return _static; }
+
+	uint32_t get_bvh_mask() const { return _bvh_mask; }
+	uint32_t get_bvh_type() const { return _bvh_type; }
+
+	ShapeOwner2DSW_Base() {
+		type = TYPE_BODY;
+		_static = true;
+
+		_bvh_mask = 0xFFFFFFFF;
+		_bvh_type = 0;
+	}
+
+	void _refresh_bvh_masks() {
+		_bvh_type = 1 << type;
+
+		// Change of behaviour. In order to fix a "bug" #17238, where
+		// users wanted static objects to collide against
+		// areas (which both have TYPE_AREA),
+		// we are changing the mask to ALWAYS collide
+		// with other objects.
+		// In order to revert, uncomment the next line.
+		//_bvh_mask = is_static() ? 0 : 0xFFFFFFFF;
```

Comment:
Ah, I see you moved the logic for pairable/unpairable here.

I'm not sure I understand why having this layer of bvh_mask and bvh_type in physics objects though. Isn't it more efficient to just use tree_id and tree_mask to replace the previous pairable/unpairable logic?

Reply:
Yes we may be able to do this for the physics (we can just return true here and it will compile out). The render tree is a bit more tricky here as it has more types etc so it may still need this mechanism, I haven't yet looked in detail at which situations we can solve with the multiple trees, as the render tree logic is a bit more complex if I remember right.

As it is, this is pretty much porting the existing behaviour, and not yet taking advantage of the tree masks. We could e.g. have first PR just port the existing behaviour, next PR changes to use the trees for this, third to change the interactions to fix the area / static bug, so we can evaluate each at a time.

Reply:
Incidentally `bvh_mask` and `bvh_type` can also alternatively be derived from the object type etc on the fly, I haven't profiled these to compare. However if we are able to use the tree masks for this, they wouldn't be necessary.

Reply:
Alright, makes sense to keep the cull check mechanism if it can be useful for render at least (and maybe in physics in the future if we encounter more corner cases where it would benefit performance).

Reply:
I've made the `user_cull_check` a null check now, so it is just relying on the tree masks.